### ISSUE TITLE
[BUG] PUNCH CARD

### ISSUE THREAD

**Issue by toritold19** [charlesbel/Microsoft-Rewards-Farmer#167]
Hello!

I'm having a problem running the MS Farmer 3.0 of Alexy... all my accounts done the first **'PUNCH CARD'** correctly (and I earn the 100 points), but when start's with the **second PUNCH CARD** stuck, after some time it skip this task and go ahead.

![image](https://user-images.githubusercontent.com/24661738/223755219-4fe287a8-152d-409c-ad38-3d0b4a738aea.png)

Something similar happen with the script when is trying to solve the **'PUZZLE'** it doesn't do it correctly. Let me know if you need another information to help me and develop any solution for the primary-script!

Thanks for your time...

**Comment by AlexyGalaxy** [comment]
I too have noticed this error and im not sure why it is occuring. 

If I find a solution I will post it here. If anyone else finds a solution please post it below. 
Thanks for making this report!

**Comment by toritold19** [comment]
Alexy let me know if we can arrange a time to run the script together to see if you detect something that we don't! I let you my Discord account: buLL#1017

**Comment by toritold19** [comment]
> I too have noticed this error and im not sure why it is occuring.
> 
> If I find a solution I will post it here. If anyone else finds a solution please post it below. Thanks for making this report!

Alexy let me know if we can arrange a time to run the script together to see if you detect something that we don't! I let you my Discord account: buLL#1017

**Comment by AlexyGalaxy** [comment]
This issue doesnt stop the script from running so I am not worried about it. If you find a solution please post it here.

My main concern is with the waituntilvisable error bc when it does occur the script cannot run. Increading the time when the function is called does not solve this.

**Comment by AlexyGalaxy** [comment]
I see what you mean the dungeons and dragons punch card starts but doenst complete the quiz. Ill look into this tomorrow.

**Comment by AlexyGalaxy** [comment]
I added timer.sleep() and that seemed to solve the issue
```
def completePunchCard(browser: WebDriver, url: str, childPromotions: dict):
    try :
        browser.get(url)
        for child in childPromotions:
            if child['complete'] == False:
                if child['promotionType'] == "urlreward":
                    browser.execute_script("document.getElementsByClassName('offer-cta')[0].click()")
                    time.sleep(2)
                    browser.switch_to.window(window_name = browser.window_handles[1])
                    time.sleep(random.randint(13, 17))
                    browser.close()
                    time.sleep(2)
                    browser.switch_to.window(window_name = browser.window_handles[0])
                    time.sleep(2)
                if child['promotionType'] == "quiz":
                    browser.execute_script("document.getElementsByClassName('offer-cta')[0].click()")
                    time.sleep(2)
                    browser.switch_to.window(window_name = browser.window_handles[1])
                    time.sleep(8)
                    try :
                        browser.find_element(By.XPATH, '//*[@id="rqStartQuiz"]').click()
                        time.sleep(1)
                    except :
                        pass
                    maxQuestions = browser.execute_script("return _w.rewardsQuizRenderInfo.maxQuestions")
                    time.sleep(3)                  
                    try :
                        numberOfOptions = browser.execute_script("return _w.rewardsQuizRenderInfo.numberOfOptions")
                        for question in range(maxQuestions):
                            if numberOfOptions == 4:
                                correctOption = browser.execute_script("return _w.rewardsQuizRenderInfo.correctAnswer")
                                for i in range(4):
                                    if browser.find_element(By.ID, "rqAnswerOption" + str(i)).get_attribute("data-option") == correctOption:
                                        browser.find_element(By.ID, "rqAnswerOption" + str(i)).click()
                                        time.sleep(5)
                                        if not waitUntilQuestionRefresh(browser):
                                            return
                                        break
                                time.sleep(5)
                    except :
                        pass
                    try: #backup code
                        counter = str(browser.find_element(By.XPATH, '//*[@id="QuestionPane0"]/div[2]').get_attribute('innerHTML'))[:-1][1:]
                        numberOfQuestions = max([int(s) for s in counter.split() if s.isdigit()])                         
                        for question in range(numberOfQuestions):
                            browser.execute_script('document.evaluate("//*[@id=\'QuestionPane' + str(question) + '\']/div[1]/div[2]/a[' + str(random.randint(1, 3)) + ']/div", document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue.click()')
                            time.sleep(5)
                            browser.find_element(By.XPATH, '//*[@id="AnswerPane' + str(question) + '"]/div[1]/div[2]/div[4]/a/div/span/input').click()
                            time.sleep(3)
                    except :
                       pass
                    time.sleep(2)
                    browser.find_element(By.XPATH, '//*[@id="rqCloseBtn"]/div/img' ).click()#quit quiz
                    time.sleep(5)
                    browser.close()
                    time.sleep(2)
                    browser.switch_to.window(window_name = browser.window_handles[0])
                    time.sleep(2)
    except:
        prRed('\n[ERROR] An Error has Occured While Trying to Complete Punch Card.\n')
        writeErr()
        FA.write('\n[ERROR] An Error has Occured While Trying to Complete Punch Card.')
        FA.close()
        time.sleep(2)
        browser.switch_to.window(window_name = browser.window_handles[0])
        time.sleep(2)
```

**Comment by toritold19** [comment]
Alexy, thanks bro... let me know when you update the script to download again the .zip!

**Comment by AlexyGalaxy** [comment]
Up now.

**Comment by charlesbel** [comment]
Fixed with v3.0 #188
